apps/emqx_rule_engine/src/emqx_rule_runtime.erl

```diff
@@ -411,13 +460,47 @@ do_handle_action(
         Result ->
             Result
     end;
-do_handle_action(RuleId, #{mod := Mod, func := Func} = Action, Selected, Envs) ->
+do_handle_action(_RuleId, #{mod := Mod, func := Func} = Action, Selected, Envs) ->
+    trace_action(Action, "call_action_function"),
     %% the function can also throw 'out_of_service'
     Args = maps:get(args, Action, []),
     Result = Mod:Func(Selected, Envs, Args),
-    inc_action_metrics(RuleId, Result),
+    TraceCtx = do_handle_action_get_trace_context(Action),
+    inc_action_metrics(TraceCtx, Result),
+    trace_action(Action, "call_action_function_result", #{result => Result}, debug),
     Result.
 
+do_handle_action_get_trace_context(Action) ->
```

Comment:
we should try NOT to build trace context (return `undefined`) if there is currently no ongoing trace (even better if there is no matching trace).

Reply:
Thanks for the suggestion. Can you explain why you think this optimization is needed? Performance reasons?

Reply:
Fixed